File: leetcode/0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (root == NULL) return res;
        
        vector<int> tmp;
        
        queue<TreeNode*> q;
        q.push(root);
        res.push_back({root->val});
        
        while (q.size() != 0)
        {
            int count = q.size();
            for (int i {0}; i < count; ++i)
            {
                TreeNode* curr = q.front();
                q.pop();
                
                if (curr->left != NULL)
                {
                    q.push(curr->left);
                    tmp.push_back(curr->left->val);
                }
                
                if (curr->right != NULL)
                {
                    q.push(curr->right);
                    tmp.push_back(curr->right->val);
                }
            }
            if (!tmp.empty())
            {
                res.push_back(tmp);
                tmp.clear();
            }
        }
        
        return res;
    }
};

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func levelOrder(_ root: TreeNode?) -> [[Int]] {
        guard let root = root else {
            return []
        }

        var queue: [TreeNode?] = []
        var res: [[Int]] = []

        queue.append(root)

        while !queue.isEmpty {
            var arr: [Int] = []

            for _ in 0...queue.count - 1 {
                let curr = queue.removeFirst()

                if let val = curr?.val {
                    arr.append(val)
                }

                if let left = curr?.left {
                    queue.append(left)
                }

                if let right = curr?.right {
                    queue.append(right)
                }
            }
            
                res.append(arr)
        }

        return res
    }
}